### DISCUSSION TITLE
Consolidated list of njsPC code suggestions for rstrouse

### DISCUSSION THREAD

**Discussion by CodeSlinger69** [tagyoureit/nodejs-poolController#821] [Nixie] [answered]
@rstrouse, given how much I messed up the pull requests and so on, let me just paste the changes here, and you can add them manually if you find them useful. I am going to exclude anything that involves supporting schedules were start time > end time, as I have learned that the design choice was to NOT allow schedules to do that. I'll also link to my detailed explanations.

### typo and logic bug when checking if a schedule is valid for today ([discussion ](https://github.com/tagyoureit/nodejs-poolController/pull/807))

controller/boards/IntelliCenterBoard.ts, line 2256
```
                if (isOn) {
                    let dt = state.time.toDate();
                    let dow = dt.getDay();
                    // Convert the dow to the bit value.
                    let sd = sys.board.valueMaps.scheduleDays.toArray().find(elem => elem.dow === dow);
-                    let dayVal = sd.bitVal || sd.val;  // The bitval allows mask overrides.
+                    //let dayVal = sd.bitVal || sd.val;  // The bitval allows mask overrides.
                    let ts = dt.getHours() * 60 + dt.getMinutes();
-                    if ((sched.scheduleDays & dayVal) > 0 && ts >= sched.startTime && ts <= sched.endTime) byte = byte | (1 << bit);
+                    if ((sched.scheduleDays & sd.bitval) > 0 && ts >= sched.startTime && ts <= sched.endTime) byte = byte | (1 << bit);
                }
```
				
controller/boards/SystemBoard.ts, line 3718
```
      // Convert the dow to the bit value.
      let sd = sys.board.valueMaps.scheduleDays.toArray().find(elem => elem.dow === dow);
-      let dayVal = sd.bitVal || sd.val;  // The bitval allows mask overrides.
+      //let dayVal = sd.bitVal || sd.val;  // The bitval allows mask overrides.
      let ts = dt.getHours() * 60 + dt.getMinutes();
      for (let i = 0; i < state.schedules.length; i++) {
        let schedIsOn: boolean;
        let ssched = state.schedules.getItemByIndex(i);
        let scirc = state.circuits.getInterfaceById(ssched.circuit);
        let mOP = sys.board.schedules.manualPriorityActive(ssched);  //sys.board.schedules.manualPriorityActiveByProxy(scirc.id);
        if (scirc.isOn && !mOP &&
-          (ssched.scheduleDays & dayVal) > 0 &&
+          (ssched.scheduleDays & sd.bitval) > 0 &&
          ts >= ssched.startTime && ts <= ssched.endTime) schedIsOn = true
        else schedIsOn = false;
```
		
controller/nixie/schedules/Schedule.ts, line 275
```
        else {
            // Convert the dow to the bit value.
            let sd = sys.board.valueMaps.scheduleDays.toArray().find(elem => elem.dow === dow);
-            let dayVal = sd.bitVal || sd.val;  // The bitval allows mask overrides.
+            //let dayVal = sd.bitVal || sd.val;  // The bitval allows mask overrides.
            // First check to see if today is one of our days.
-            if ((this.schedule.scheduleDays & dayVal) === 0) return false;
+            if ((this.schedule.scheduleDays & sd.bitval) === 0) return false;
        }
```
 
### fixing nocturnal cooling logic ([discussion ](https://github.com/tagyoureit/nodejs-poolController/discussions/808#discussioncomment-6347827))

controller/boards/SystemBoard.ts, line 4273
```
	else if (heater.coolingEnabled && body.temp > cfgBody.coolSetpoint && state.heliotrope.isNight &&
-       state.temps.solar > body.temp + (hstate.isOn ? heater.stopTempDelta : heater.startTempDelta)) {
+	state.temps.solar < body.temp - (hstate.isOn ? heater.stopTempDelta : heater.startTempDelta)) {
		isOn = true;
		body.heatStatus = sys.board.valueMaps.heatStatus.getValue('cooling');
		isHeating = true;
		isCooling = true;
	}
```

### fixing isNight? logic, ([discussion ](https://github.com/tagyoureit/nodejs-poolController/pull/818))
Note: From the discussion , it looks like you solved this a different way, so this is moot. I was able to test your new code and it works from all the tests that I did. FYI, line 177 is reached when njsPC is started up between sunset and midnight.

controller/Constants.ts, line 128
```
-    public get isNight(): boolean {
-        let times = this.calculatedTimes;
-        if (this.isValid) {
-            let time = new Date().getTime();
-            if (time >= times.sunset.getTime() && time < times.sunrise.getTime()) return true;
-        }
-        return false;
-    }
+    public get isNight(): boolean {
+        if (!this.isValid) return false;
+
+        let times = this.calculatedTimes;
+        let sunriseMins = times.sunrise.getHours() * 60 + times.sunrise.getMinutes();
+        let sunsetMins = times.sunset.getHours() * 60 + times.sunset.getMinutes();
+        let now = new Date()
+        let nowMins = now.getHours() * 60 + now.getMinutes();
+
+        if (nowMins > sunriseMins && nowMins <= sunsetMins) return false;
+        else return true;
+    }
```

### fix for circuits with schedules that overlap in a polling window ([discussion](https://github.com/tagyoureit/nodejs-poolController/pull/813))
Note: In that discussion you mentioned that this is not the correct way to solve this problem.

controller/nixie/schedules/Schedule.ts, line 214
```
            else if (shouldBeOn && this.running) {
+                // Check to see if circuit is on, if not, turn it on
+                if (!cstate.isOn) ctx.setCircuit(circuit.id, true);
+
                // With mOP, we need to see if the schedule will come back into play and also set the circut
                if (this.suspended && cstate.isOn) {
                    if (sys.general.options.manualPriority) {
```

**Comment by rstrouse** [comment, marked as answer]
Funny, I found this post after I fixed it all.  There was a bit of a different approach taken to the schedules and isNight to accommodate the sunrise changes for the following day.

**Reply by CodeSlinger69** [reply]
Robert, thank you so much for your patience with me as I work through getting njsPC working for me, and also with my github screwups. I do have a feeling that I am the first one to try some of these things in Nixie. Such as automating nocturnal cooling and having different schedules on different days for the same circuit.

On the github thing, if I do see any potential opportunities to commit code in the future, I'll just post them in a thread like this in code tags, and avoid all that git/github complexity.

But, things are looking awesome now. I'll be testing furiously, and hopefully getting nocturnal cooling working. Its very close...
